docs: streamline 'timeout' / 'retry' docs in docstrings

tseaver opened the pull request:
* Add 'requests' intersphinx refs.
* Add narrative docs for timeouts and retries.
* Replace boilerplate docs for 'timeout' and 'retry' in API-invoking methods with links to narrative docs.

Closes #455

tseaver committed:
docs: add 'requests' intersphinx refs

tseaver committed:
docs: add narrative docs for timeouts and retries

Include API docs for the 'retry' module, as well, to unbreak links.

tseaver committed:
docs: streamline 'timeout' / 'retry docstrngs

Replace boilerplate with links to new narrative docs.

Closes #455.

tseaver commented on owlbot.py:
```diff
@@ -38,4 +38,17 @@
+    "requests": ("https://docs.python-requests.org/en/master/", None),
+""",
+)
+
```
Hmm, owlbot seems not to be applying this change:  036499972586c97ca79fc9796fbc56b1bbd510a3 removes the entry for `"requests"`.  @parthea, @busunkim96 can you help?

busunkim96 replied:
It looks like you can pass dict to `py_library` with additional intersphinx mappings:

https://github.com/googleapis/synthtool/blob/2430f8d90ed8a508e8422a3a7191e656d5a6bf53/synthtool/gcp/templates/python_library/docs/conf.py.j2#L370-L374

```py
common.py_library(
    ...,
    intersphinx_dependencies={
        "requests": "https://docs.python-requests.org/en/master/"
    }
)
```

Not sure why the current replacement is getting undone though.

tseaver replied:
I needed to escape the parens in the text-to-be-replaced. :)

tseaver committed:
fix: appease regex matching idols

tseaver committed:
fix: fail owlbot if replacement misses

tseaver committed:
fix: add 'requests' intersphinx dep via library API

Rather than hacking it in via replacement.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

tritone commented on docs/retry_timeout.rst:
```diff
@@ -0,0 +1,114 @@
+Methods which invoke API methods may fail for a number of reasons, some of
+which represent "transient" conditions, and thus can be retried
+automatically.  The library tries to provide a sensible default retry policy
+for each method, base on its semantics.
```
I would add a little more color on the default retry policy at the top here. Generally, the strategy is to retry any idempotent operation (some calls are always idempotent while others require a precondition to qualify) on any transient error.

tseaver replied:
4278187

tritone commented on docs/retry_timeout.rst:
```diff
@@ -0,0 +1,114 @@
+   from google.api_core.retry import Retry
+
+   def is_retryable(exc):
+       ... # your retriable exception types here
```
Maybe add an example condition here for clarity.

tseaver replied:
4278187

tritone reviewed (commented):
Thanks for adding this! A couple small suggestions.

tritone commented:
@andrewsg @cojenco PTAL

tseaver committed:
docs: add docstrings for default conditional policies.

tseaver committed:
docs: flesh out retry docs

- Explain semantics of default policies, and where used.

- Provide an example of a predicate function for a conditional policy.

Per review comments.

tritone commented on docs/retry_timeout.rst:
```diff
@@ -58,9 +58,31 @@ Configuring Retries
+  retries any API request which returns a "transient" error.
+
+- For API requests which are idempotent only if the bucket or blob has
+  the same "metageneration", the library uses its
```
Should be "generation" here

tseaver replied:
2b071b0

tritone commented on docs/retry_timeout.rst:
```diff
@@ -58,9 +58,31 @@ Configuring Retries
+  :data:`~google.cloud.storage.retry.DEFAULT_RETRY_IF_ETAG_IN_JSON`
+  policy, which retries API requests which returns a "transient" error,
+  but only if the original request includes an ``ETAG`` in its payload.
+
```
There are also a few APIs which are always non-idempotent; these are never retried.

tseaver replied:
2b071b0

tritone reviewed (commented):
Looks great!

tseaver committed:
docs: tweak

Per review comments.

cojenco commented on docs/retry_timeout.rst:
```diff
@@ -0,0 +1,147 @@
+  :data:`~google.cloud.storage.retry.DEFAULT_RETRY` policy, which
+  retries any API request which returns a "transient" error.
+
+- For API requests which are idempotent only if the bucket or blob has
```
We can remove `bucket` here. Since buckets have no payload data, [buckets have no generation numbers](https://cloud.google.com/storage/docs/generations-preconditions).  `if_generation_match` should only apply to blobs.

tseaver replied:
4bb0001

cojenco commented on docs/retry_timeout.rst:
```diff
@@ -0,0 +1,147 @@
+  retries any API request which returns a "transient" error.
+
+- For API requests which are idempotent only if the bucket or blob has
+  the same "generation", the library uses its
```
Since we are streamlining the retry strategy reference to one place, should we include here a descriptive use case: `Setting to 0 makes the operation succeed only if there are no live versions of the blob.`

tseaver replied:
That language would actually be documentation for the `if_generation_match` argument, rather than the retry policy:  I've proposed doing a similar streamline for those arguments in #460.

tseaver committed:
docs: tweak

Further review comments.

cojenco commented on docs/retry_timeout.rst:
```diff
@@ -0,0 +1,147 @@
+
+See also:
+
+  :ref:`Timeouts in requests <requests:timeouts>`
```
Looks great, thanks! @andrewsg and our TW, Karen, also wrote [documentation on Retry Strategies specifically for the Python client library](https://cloud.google.com/storage/docs/retry-strategy#python). Maybe we could link it as a reference too.

tseaver replied:
1a56e89

cojenco reviewed (approved):
LGTM, thanks @tseaver for streamlining this!

tseaver committed:
docs: add ref to back-end retry strategy docs

Per review.
